### PR TITLE
Extract PortRangeSocketBinder from Server.

### PR DESCRIPTION
**Pull request by mneedham**
Implementation now also supports inverted port ranges;
e.g. localhost:6010-6000

Authors: me and @apcj

### CODE REVIEW

File: enterprise/com/src/main/java/org/neo4j/com/PortChecker.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+package org.neo4j.com;
+
+public class PortChecker
+{
+
+    public boolean isPortFree( int port )
+    {
+        return false;
+    }
+}
```
Comment (davidegrohmann) [inline enterprise/com/src/main/java/org/neo4j/com/PortChecker.java]: This looks unneeded and weird, can this class be removed?

File: enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import org.jboss.netty.channel.ChannelException;
+
+import java.net.InetSocketAddress;
+
+import org.neo4j.helpers.HostnamePort;
+
+public class PortRangeSocketBinder
+{
+    private ServerBootstrap bootstrap;
+    private static final String ALL_INTERFACES_ADDRESS = "0.0.0.0";
```
Comment (davidegrohmann) [inline enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java:33]: Move this constant to `HostnamePort`?

File: enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                socketAddress = new InetSocketAddress( host, port );
+            }
+            try
+            {
+                channel = bootstrap.bind( socketAddress );
+                return new Connection( socketAddress, channel );
+            }
+            catch ( ChannelException e )
+            {
+                lastException = e;
```
Comment (davidegrohmann) [inline enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java:72]: Using `addSuppressed` might help here?

**Comment by davidegrohmann** [comment]
@mneedham Let's merge this in when my comments are addressed.

File: enterprise/com/src/main/java/org/neo4j/com/PortChecker.java (hunk 1 above)
Comment (mneedham) [inline enterprise/com/src/main/java/org/neo4j/com/PortChecker.java]: Yup. That was committed by mistake. Will remove

File: enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java (hunk 3 above)
Comment (mneedham) [inline enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java:72]: My IntelliJ isn't flagging any error - which line do we need the `addSuppressed` on?

File: enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java (hunk 2 above)
Comment (mneedham) [inline enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java:33]: IIRC it was there originally but the only place where it's used is in this class hence we moved it over here

File: enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java (hunk 3 above)
Comment (davidegrohmann) [inline enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java:72]: Sorry my bad, I didn't explain it properly. I was trying to suggest something like:

```
if (lastException != null)
{
   lastException = e
}
else
{
  e.addSuppressed(lastException);
  lastException = e;
} 
```

In order not to swallow all the exceptions.

File: enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java (hunk 2 above)
Comment (davidegrohmann) [inline enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java:33]: Fair enough

File: enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java (hunk 3 above)
Comment (mneedham) [inline enterprise/com/src/main/java/org/neo4j/com/PortRangeSocketBinder.java:72]: Good idea. Have added that logic in + a test to make sure it returns the appropriate suppressed exceptions

**Pull request merged**